### ISSUE TITLE
Float128 calling convention and alignment

### ISSUE THREAD

**Issue by simonbyrne** [JuliaMath/Quadmath.jl#4]
At the moment we fake the calling convention on Mac and Linux by reinterpreting as `NTuple{2,VecElement{Float64}}`. Unfortunately this doesn't seem to work on Windows.

Upstream:
- https://github.com/JuliaLang/julia/pull/22649
- https://github.com/JuliaLang/julia/issues/20415
- https://github.com/JuliaLang/julia/pull/21959

Also mentioned in 
- https://github.com/JuliaLang/julia/issues/21216

**Comment by simonbyrne** [comment]
cc: @vchuravy @vjtnash @yuyichao

**Comment by yuyichao** [comment]
Just a thought, have you try to add a `x86_vectorcallcc` on the `ccall`? Tht seems to produce the correct calling convention assuming you only have `fp128`.

**Comment by simonbyrne** [comment]
how do I do that? is that an undocumented calling convention?

**Comment by simonbyrne** [comment]
Ah, we don't seem to support it. Should I open an issue?

**Comment by yuyichao** [comment]
I don't think our `ccall` supports it and I'm not sure where it's documented in LLVM but you can use [this trick](https://github.com/JuliaLang/julia/pull/21589/files#diff-e5503124227910a9007ce5c88dca7e7aR279)

**Comment by yuyichao** [comment]
Note that I think you won't be able to use it even if we support it since supporting it in `ccall` would probably need to come with automatic name mangling support...

**Comment by simonbyrne** [comment]
I see, thanks.

**Comment by RalphAS** [comment]
I tried some experiments following @yuyichao 's suggestion, for example
```julia
function baz(x::Float128, y::Float128)
    r = Base.llvmcall("""%f = inttoptr i64 %2 to <2 x double> (<2 x double>, <2 x double>)*
                    %vv = call x86_vectorcallcc <2 x double> %f(<2 x double> %0, <2 x double> %1)
                    ret <2 x double> %vv""",
                 Cfloat128, Tuple{Cfloat128,Cfloat128,Ptr{Cvoid}},
                 x.data, y.data, cglobal((:__addtf3,quadoplib)))
    return Float128(r)
end
```
Replacing methods in Quadmath with patterns like this works fine for Linux and OSX.  It runs on Windows (no segfaults) but produces incorrect answers, apparently because the Windows version of libgcc_s doesn't return Float128 results in xmm0 as expected from the ABI. (AFAICT the value of xmm0 is preserved across the call.)

**Comment by yuyichao** [comment]
Come to think about it, depending on if LLVM and GCC agrees on the calling convention, if you are already using llvmcall, you can probably directly use fp128 in it ;-p....

Other than that, I have no idea what calling convention gcc uses..... It is worth noting that GCC does **NOT** use the same vector calling convention as clang or msvc. (I didn't realize you are calling gcc compiled library....) I just finished dealing with a similar issue with vector calls on windows so I'm pretty positive on that....... You might have to check the assembly code to figure out what calling convention gcc is using..... =(

**Comment by simonbyrne** [comment]
> (I didn't realize you are calling gcc compiled library....)

we're calling into libquadmath, which is bundled as part of the gcc runtime (which we ship with Julia apparently)

**Comment by simonbyrne** [comment]
> Replacing methods in Quadmath with patterns like this works fine for Linux and OSX. It runs on Windows (no segfaults) but produces incorrect answers, apparently because the Windows version of libgcc_s doesn't return Float128 results in xmm0 as expected from the ABI. (AFAICT the value of xmm0 is preserved across the call.)

Not sure if it makes a difference, but did you change how `Cfloat128` is defined on Windows to be the same as linux and mac?

**Comment by RalphAS** [comment]
Yes, I changed the `Float128`/`Cfloat128` structure for Windows to resemble the others.

Note that the Quadmath conversions, comparisons, and arithmetic are passed to libgcc_s (not libquadmath) in Windows and Linux.

I was surprised to learn that LLVM `fp128` instructions are also compiled into libgcc_s calls (according to `code_native`). My original thought was that "direct" `fp128` IR might help the Windows issue, but that turned out to be pointless (other than helping me to learn LLVM).

FWIW, I found that IR using `fp128` is rather fragile, in that it can cause LLVM to crash Julia or go into a coma (presumably a recurrent loop). That's why the above example just uses `<2 x double>`.
